pkg/dockerloginner/dockerloginner.go

```diff
@@ -0,0 +1,133 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockerloginner
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/nuclio/nuclio/pkg/dockerclient"
+	"github.com/nuclio/nuclio/pkg/errors"
+
+	"github.com/nuclio/nuclio-sdk"
+)
+
+// DockerLoginner initializes docker client credentials
+type DockerLoginner struct {
+	logger       nuclio.Logger
+	dockerClient dockerclient.Client
+}
+
+func NewDockerLoginner(parentLogger nuclio.Logger,
+	dockerClient dockerclient.Client) (*DockerLoginner, error) {
+
+	return &DockerLoginner{
+		logger:       parentLogger.GetChild("loginner"),
+		dockerClient: dockerClient,
+	}, nil
+}
+
+func (dl *DockerLoginner) LoginFromDir(keyDir string) error {
+	dockerKeyDirFileInfos, err := ioutil.ReadDir(keyDir)
+	if err != nil {
+		return errors.Wrap(err, "Failed to read docker key directory")
+	}
+
+	for _, dockerKeyDirFileInfo := range dockerKeyDirFileInfos {
+		if dockerKeyDirFileInfo.IsDir() {
+			continue
+		}
+
+		dockerKeyPath := path.Join(keyDir, dockerKeyDirFileInfo.Name())
+
+		dockerKeyDirFile, err := os.Open(dockerKeyPath)
+		if err != nil {
+			dl.logger.WarnWith("Failed to open docker key file",
+				"err", err.Error(),
+				"path", dockerKeyDirFileInfo.Name())
+
+			continue
+		}
+
+		// get the URL and username
+		username, url, err := dl.getUserAndURLFromKeyPath(dockerKeyDirFile.Name())
+		if err != nil {
+			dl.logger.WarnWith("Failed to get user / url from path",
+				"err", err.Error(),
+				"path", dockerKeyDirFileInfo.Name())
+
+			continue
+		}
+
+		// read the password
+		password, err := ioutil.ReadAll(dockerKeyDirFile)
+		if err != nil {
+			dl.logger.WarnWith("Failed to read docker key",
+				"err", err.Error(),
+				"path", dockerKeyDirFileInfo.Name())
+
+			continue
+		}
+
+		dl.logger.InfoWith("Logging in to registry",
+			"path", dockerKeyDirFile.Name(),
+			"username", username,
+			"passwordLen", len(password),
+			"url", url)
+
+		// try to login
+		err = dl.dockerClient.LogIn(&dockerclient.LogInOptions{
+			Username: username,
+			Password: string(password),
+			URL:      "https://" + url,
+		})
+
+		if err != nil {
+			dl.logger.WarnWith("Failed to log in to docker", "err", err.Error())
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (dl *DockerLoginner) getUserAndURLFromKeyPath(keyPath string) (string, string, error) {
+	dockerKeyBase := path.Base(keyPath)
+	dockerKeyExt := path.Ext(dockerKeyBase)
+
+	// get just the file name
+	dockerKeyFileWithoutExt := dockerKeyBase[:len(dockerKeyBase)-len(dockerKeyExt)]
+
+	// expect user@url in the ext
+	userAndURL := strings.Split(dockerKeyFileWithoutExt, "---")
+	if len(userAndURL) != 2 {
```

Comment:
Refactor to `validateUserAndURL` function?

Reply:
I think it's small enough, kinda weird to break it down further